### PR TITLE
fix fieldref assignment to not assume mutable object	

### PR DESCRIPTION
**Pull request by colinsurprenant**
this is related to problem described in elastic/logstash#4140 and also relates to elastic/logstash#4191
we cannot assume anymore that field reference getter returns a mutable object.

tracked in elastic/logstash#4264

sorry for the whitespaces diffs :(

### CODE REVIEW

File: lib/logstash/filters/grok.rb [hunk 1]
```diff
@@ -288,21 +288,28 @@ def filter(event)
...
         @tag_on_failure.each do |tag|
-          event["tags"] ||= []
-          event["tags"] << tag unless event["tags"].include?(tag)
+          # do not replace the code below with:
+          #   event["tags"] ||= []
+          #   event["tags"] << tag unless event["tags"].include?(tag)
+          # this assumes implementation specific feature of returning a mutable object
+          # from a field ref which should not be assumed and will change in the future.
+
+          tags = event["tags"] || []
```
Comment (jsvd) [inline lib/logstash/filters/grok.rb]: can we just replace this with `event.tag(tag)` and change the [default implementation of tag](https://github.com/elastic/logstash/blob/master/logstash-core-event/lib/logstash/event.rb#L201-L205) to do this?

I think this cleanup fits the whole refactoring you're doing :)

Comment (colinsurprenant) [inline lib/logstash/filters/grok.rb]: oh yes, good catch, I don't know why I did not correctly clean this. is can be replaced with a simple: 
```ruby
event.tag(tag)
```

**Comment by colinsurprenant** [comment]
@jsvd followup review request please!

File: lib/logstash/filters/grok.rb [hunk 2]
```diff
@@ -288,21 +288,26 @@ def filter(event)
...
+
       if matched
         filter_matched(event)
       else
         # Tag this event if we can't parse it. We can use this later to
         # reparse+reindex logs if we improve the patterns given.
         @tag_on_failure.each do |tag|
-          event["tags"] ||= []
-          event["tags"] << tag unless event["tags"].include?(tag)
+          # do not replace the code below with:
```
Comment (jsvd) [inline lib/logstash/filters/grok.rb]: I think it's safe to remove this warning, it makes sense in the example below where a developer will instinctively try to optimize/dry/refactor code, here I doubt someone would replace `event.tag(tag)` with 2 lines of more complex code

**Comment by jsvd** [comment]
minor comment otherwise LGTM

File: lib/logstash/filters/grok.rb (hunk 2 above)
Comment (colinsurprenant) [inline lib/logstash/filters/grok.rb]: good point :P  did not really pay attention to the comment.

**Comment by colinsurprenant** [comment]
thanks, will fix comment and merge!

File: lib/logstash/filters/grok.rb (hunk 2 above)
Comment (colinsurprenant) [inline lib/logstash/filters/grok.rb]: done!

**Comment by colinsurprenant** [comment]
pushed version 2.0.3

**Pull request merged**